File: sdks/lua/adriane.lua
local ffi = require("ffi")

ffi.cdef[[
typedef struct AdrianeResult {
  int code;
  char *value;
  char *error;
} AdrianeResult;

typedef int (*AdrianeStringCallback)(const char *payload_json, void *user_data, const char **value, const char **error);
typedef void (*AdrianeEventCallback)(const char *payload_json, void *user_data);

typedef struct AdrianeCallbacks {
  void *user_data;
  AdrianeStringCallback on_node;
  AdrianeStringCallback on_condition;
  AdrianeEventCallback on_event;
} AdrianeCallbacks;

char *adriane_engine_version(void);
AdrianeResult adriane_validate_graph_json(const char *definition_json);
AdrianeResult adriane_compile_graph_yaml_json(const char *yaml);
AdrianeResult adriane_available_providers_json(void);
AdrianeResult adriane_resolve_model_json(const char *tier, const char *available_json, const char *override_json);
AdrianeResult adriane_list_components_json(void);
AdrianeResult adriane_list_prebuilt_json(void);
AdrianeResult adriane_run_component_json(const char *kind, const char *params_json, const char *channels_json);
AdrianeResult adriane_run_prebuilt_json(const char *name, const char *input_json, const char *options_json);
AdrianeResult adriane_engine_run_json(const char *spec_json, AdrianeCallbacks callbacks);
AdrianeResult adriane_engine_resume_json(const char *spec_json, AdrianeCallbacks callbacks);
AdrianeResult adriane_engine_approve_and_resume_json(const char *spec_json, AdrianeCallbacks callbacks);
AdrianeResult adriane_engine_signal_json(const char *spec_json, const char *signal_name, const char *payload_json, AdrianeCallbacks callbacks);
AdrianeResult adriane_engine_replay_json(const char *spec_json, const char *checkpoint_id, AdrianeCallbacks callbacks);
void adriane_string_free(char *ptr);
void adriane_result_free(AdrianeResult result);
]]

local M = {}
local lib = nil

local function default_library_name()
  if ffi.os == "OSX" then
    return "libadriane_c_api.dylib"
  end
  if ffi.os == "Windows" then
    return "adriane_c_api.dll"
  end
  return "libadriane_c_api.so"
end

function M.load(path)
  lib = ffi.load(path or os.getenv("ADRIANE_C_API_LIB") or default_library_name())
  return M
end

local function engine()
  if lib == nil then
    M.load()
  end
  return lib
end

local function take_string(ptr)
  if ptr == nil or ptr == ffi.NULL then
    return nil
  end
  local value = ffi.string(ptr)
  engine().adriane_string_free(ptr)
  return value
end

local function unwrap(result)
  local native = engine()
  if result.code == 0 then
    local value = ffi.string(result.value)
    native.adriane_result_free(result)
    return value
  end

  local message = "Adriane C API error " .. tostring(result.code)
  if result.error ~= nil and result.error ~= ffi.NULL then
    message = ffi.string(result.error)
  end
  native.adriane_result_free(result)
  error(message, 2)
end

function M.engine_version()
  local ptr = engine().adriane_engine_version()
  if ptr == nil or ptr == ffi.NULL then
    error("adriane_engine_version returned null", 2)
  end
  return take_string(ptr)
end

function M.validate_graph_json(definition_json)
  return unwrap(engine().adriane_validate_graph_json(definition_json))
end

function M.compile_graph_yaml_json(yaml)
  return unwrap(engine().adriane_compile_graph_yaml_json(yaml))
end

function M.available_providers_json()
  return unwrap(engine().adriane_available_providers_json())
end

function M.resolve_model_json(tier, available_json, override_json)
  return unwrap(engine().adriane_resolve_model_json(tier, available_json, override_json))
end

function M.list_components_json()
  return unwrap(engine().adriane_list_components_json())
end

function M.list_prebuilt_json()
  return unwrap(engine().adriane_list_prebuilt_json())
end

function M.run_component_json(kind, params_json, channels_json)
  return unwrap(engine().adriane_run_component_json(kind, params_json, channels_json))
end

function M.run_prebuilt_json(name, input_json, options_json)
  return unwrap(engine().adriane_run_prebuilt_json(name, input_json, options_json))
end

function M.engine_run_json(spec_json, callbacks)
  return unwrap(engine().adriane_engine_run_json(spec_json, callbacks))
end

function M.engine_resume_json(spec_json, callbacks)
  return unwrap(engine().adriane_engine_resume_json(spec_json, callbacks))
end

function M.engine_approve_and_resume_json(spec_json, callbacks)
  return unwrap(engine().adriane_engine_approve_and_resume_json(spec_json, callbacks))
end

function M.engine_signal_json(spec_json, signal_name, payload_json, callbacks)
  return unwrap(engine().adriane_engine_signal_json(spec_json, signal_name, payload_json, callbacks))
end

function M.engine_replay_json(spec_json, checkpoint_id, callbacks)
  return unwrap(engine().adriane_engine_replay_json(spec_json, checkpoint_id, callbacks))
end

return M

File: sdks/php/src/Adriane.php
<?php

final class Adriane
{
    private \FFI $ffi;

    private function __construct(\FFI $ffi)
    {
        $this->ffi = $ffi;
    }

    public static function load(?string $library = null): self
    {
        $library = $library ?: self::defaultLibraryName();
        $cdef = <<<'CDEF'
typedef struct AdrianeResult {
  int code;
  char *value;
  char *error;
} AdrianeResult;

char *adriane_engine_version(void);
AdrianeResult adriane_validate_graph_json(const char *definition_json);
AdrianeResult adriane_compile_graph_yaml_json(const char *yaml);
AdrianeResult adriane_available_providers_json(void);
AdrianeResult adriane_resolve_model_json(const char *tier, const char *available_json, const char *override_json);
AdrianeResult adriane_list_components_json(void);
AdrianeResult adriane_list_prebuilt_json(void);
AdrianeResult adriane_run_component_json(const char *kind, const char *params_json, const char *channels_json);
AdrianeResult adriane_run_prebuilt_json(const char *name, const char *input_json, const char *options_json);
void adriane_string_free(char *ptr);
void adriane_result_free(AdrianeResult result);
CDEF;
        return new self(\FFI::cdef($cdef, $library));
    }

    public function engineVersion(): string
    {
        $ptr = $this->ffi->adriane_engine_version();
        if ($ptr === null) {
            throw new \RuntimeException("adriane_engine_version returned null");
        }

        try {
            return \FFI::string($ptr);
        } finally {
            $this->ffi->adriane_string_free($ptr);
        }
    }

    public function validateGraphJson(string $definitionJson): string
    {
        return $this->unwrap($this->ffi->adriane_validate_graph_json($definitionJson));
    }

    public function compileGraphYamlJson(string $yaml): string
    {
        return $this->unwrap($this->ffi->adriane_compile_graph_yaml_json($yaml));
    }

    public function availableProvidersJson(): string
    {
        return $this->unwrap($this->ffi->adriane_available_providers_json());
    }

    public function resolveModelJson(string $tier, ?string $availableJson = null, ?string $overrideJson = null): string
    {
        return $this->unwrap($this->ffi->adriane_resolve_model_json($tier, $availableJson, $overrideJson));
    }

    public function listComponentsJson(): string
    {
        return $this->unwrap($this->ffi->adriane_list_components_json());
    }

    public function listPrebuiltJson(): string
    {
        return $this->unwrap($this->ffi->adriane_list_prebuilt_json());
    }

    public function runComponentJson(string $kind, string $paramsJson, string $channelsJson): string
    {
        return $this->unwrap($this->ffi->adriane_run_component_json($kind, $paramsJson, $channelsJson));
    }

    public function runPrebuiltJson(string $name, string $inputJson, ?string $optionsJson = null): string
    {
        return $this->unwrap($this->ffi->adriane_run_prebuilt_json($name, $inputJson, $optionsJson));
    }

    private function unwrap(\FFI\CData $result): string
    {
        try {
            if ($result->code === 0) {
                return \FFI::string($result->value);
            }

            $message = $result->error === null
                ? "Adriane C API error {$result->code}"
                : \FFI::string($result->error);
            throw new \RuntimeException($message);
        } finally {
            $this->ffi->adriane_result_free($result);
        }
    }

    private static function defaultLibraryName(): string
    {
        return match (PHP_OS_FAMILY) {
            "Darwin" => "libadriane_c_api.dylib",
            "Windows" => "adriane_c_api.dll",
            default => "libadriane_c_api.so",
        };
    }
}
